### ISSUE TITLE
Elide identity mappings

### ISSUE THREAD

**Issue by wemrysi** [quasar-analytics/quasar#1569]
A request to elide `Map((), SrcHole)` nodes from `QScript` as seen in the plan for the following query

`select * from /ml8/census/some-zips`
```
Logical Plan:
Let('tmp0)
├─ Read(/ml8/census/some-zips)
╰─ Let('tmp1)
   ├─ Invoke(SQUASH)
   │  ╰─ Free('tmp0)
   ╰─ Free('tmp1)

Optimized:
Invoke((LIMIT))
├─ Invoke(SQUASH)
│  ╰─ Read(/ml8/census/some-zips)
╰─ Constant(Int(11))

Typechecked:
Invoke((LIMIT))
├─ Invoke(SQUASH)
│  ╰─ Read(/ml8/census/some-zips)
╰─ Constant(Int(11))

QScript:
List
├─ Unreferenced
╰─ Subset((),LeftShift(Read(/census/some-zips),SrcHole,RightSide),Take,Map(Unreferenced,Constant(Int(11))))
   ╰─ ExternallyManaged(Extern)

QScript (ShiftRead):
List
├─ Unreferenced
╰─ Subset((),Map(ShiftedRead(/census/some-zips, ExcludeId),SrcHole),Take,Map(Unreferenced,Constant(Int(11))))
   ╰─ ExternallyManaged(Extern)

QScript (Optimized):
List
├─ Unreferenced
├─ Subset((),ShiftedRead(/census/some-zips, ExcludeId),Take,Map(Unreferenced,Constant(Int(11))))
│  ╰─ ExternallyManaged(Extern)
╰─ Map((),SrcHole)
   ╰─ ExternallyManaged(Extern)
```

**Comment by sellout** [comment]
@wemrysi This is a normalization that exists. I can make the change, but I’m not set up to run tests against MarkLogic. If you want to make the change yourself, here it is:

```diff
modified   marklogic/src/main/scala/quasar/physical/marklogic/fs/queryfile.scala
@@ -65,6 +65,7 @@ object queryfile {
       // TODO[scalaz]: Shadow the scalaz.Monad.monadMTMAB SI-2712 workaround
       import WriterT.writerTMonad
       val rewrite = new Rewrite[Fix]
+      val C = Coalesce[Fix, MLQScript, MLQScript]
 
       def phase(main: MainModule): PhaseResults =
         Vector(PhaseResult.detail("XQuery", main.render))
@@ -85,7 +86,12 @@ object queryfile {
         shifted =  shiftRead[Fix](qs)
         _       <- MonadTell[QPlan, PhaseResults].tell(Vector(
                      PhaseResult.tree("QScript (ShiftRead)", shifted.cata(linearize).reverse)))
-        optmzed =  shifted.transCata(rewrite.optimize(reflNT))
+        optmzed =  shifted
+                     .transAna(
+                       repeatedly(C.coalesceQC[MLQScript](idPrism)) ⋙
+                       repeatedly(C.coalesceTJ[MLQScript](idPrism.get)) ⋙
+                       repeatedly(C.coalesceSR[MLQScript](idPrism)))
+                     .transCata(rewrite.optimize(reflNT))
         _       <- MonadTell[QPlan, PhaseResults].tell(Vector(
                      PhaseResult.tree("QScript (Optimized)", optmzed.cata(linearize).reverse)))
         mod     <- plan(optmzed).leftMap(mlerr => mlerr match {
```

I can submit it if you’d prefer, though.

**Comment by wemrysi** [comment]
@sellout Ok, I can look into it next week. As an aside, if you have `docker` and `jq` installed, you should be able to run `scripts/installMarkLogic` to get a running container.

**Comment by alissapajer** [comment]
Closed by the merge of #1582
